nvtx.hpp

```diff
@@ -0,0 +1,1810 @@
+ *
+ * \subsection PAYLOAD Payload
+ *
+ * Allows associating a user-defined numerical value with an event.
+ *
+ * ```
+ * nvtx:: EventAttributes attr{nvtx::Payload{42}}; // Constructs a Payload from
+ *                                                 // the `int32_t` value 42
+ * ```
+ *
+ *
+ * \section EXAMPLE Example
+ *
+ * Putting it all together:
+ * \code{.cpp}
+ * // Define a custom domain tag type
+ * struct my_domain{ static constexpr char const* name{"my domain"}; };
+ *
+ * // Define a named category tag type
+ * struct my_category{
+ *    static constexpr char const* name{"my category"};
+ *    static constexpr uint32_t id{42};
+ * };
+ *
+ * // Define a registered message tag type
+ * struct my_message{ static constexpr char const* message{"my message"}; };
+ *
+ * // For convenience, use aliases for domain scoped objects
+ * using my_thread_range = nvtx::domain_thread_range<my_domain>;
+ * using my_registered_message = nvtx::RegisteredMessage<my_domain>;
+ * using my_named_category = nvtx::NamedCategory<my_domain>;
+ *
+ *
+ * // Default values for all attributes
+ * nvtx::EventAttributes attr{}
+ * my_thread_range r0{attr};
+ *
+ * // Custom (unregistered) message, and unnamed Category
+ * nvtx::EventAttributes attr1{"message", nvtx::Category{2}};
+ * my_thread_range r1{attr1};
+ *
+ * // Alternatively, pass arguments of `EventAttributes` ctor directly to `my_thread_range`
+ * my_thread_range r2{"message", nvtx::Category{2}};
+ *
+ * // construct on first use a registered message
+ * auto msg = my_registered_message::get<my_message>();
+ *
+ * // construct on first use a named category
+ * auto category = my_named_category::get<my_category>();
+ *
+ * // Use registered message and named category
+ * my_thread_range r3{msg, category, nvtx::RGB{127, 255, 0}, nvtx::Payload{42}};
+ *
+ * // Any number of arguments in any order
+ * my_thread_range r{nvtx::RGB{127, 255,0}, msg};
+ *
+ * \endcode
+ * \section MACROS Convenience Macros
+ *
+ * Oftentimes users want to quickly and easily add NVTX ranges to their library
+ * or application to aid in profiling and optimization.
+ *
+ * A convenient way to do this is to use the \ref NVTX_FUNC_RANGE and
+ * \ref NVTX_FUNC_RANGE_IN macros. These macros take care of constructing an
+ * `nvtx::domain_thread_range` with the name of the enclosing function as the
+ * range's message.
+ *
+ * \code{.cpp}
+ * void some_function(){
+ *    // Automatically generates an NVTX range for the duration of the function
+ *    // using "some_function" as the event's message.
+ *    NVTX_FUNC_RANGE();
+ * }
+ * \endcode
+ *
+ */
+
+/**
+ * @brief Enables the use of constexpr when support for C++14 relaxed constexpr
+ * is present.
+ *
+ * Initializing a legacy-C (i.e., no constructor) union member requires
+ * initializing in the constructor body. Non-empty constexpr constructors
+ * require C++14 relaxed constexpr.
+ *
+ */
+#if __cpp_constexpr >= 201304L
+#define NVTX_RELAXED_CONSTEXPR constexpr
```

Comment:
Does NVTX++ require building .cu files with `--expt-relaxed-constexpr`?  Should define what "relaxed constexpr is present" means here for non-CUDA experts.

Reply:
Nope, this is a different kind of relaxed constexpr. Doesn't have anything to do with NVCC or device code. https://en.wikipedia.org/wiki/C%2B%2B14#Relaxed_constexpr_restrictions